### PR TITLE
Feature/scrutinizer optimizations

### PR DESCRIPTION
**Pull request by p4u**
see commit message

### CODE REVIEW

File: chain/ethevents/events.go [hunk 1]
```diff
@@ -150,60 +150,66 @@ func NewEthEvents(
...
+			log.Infof("process %x results already added on the Vochain, skipping",
+				results.ProcessID)
 			return
 		}
 	case models.ProcessStatus_ENDED:
 		break
 	default:
-		log.Infof("invalid process %x status %s for setting the results, skipping", results.ProcessId, vocProcessData.Status)
+		log.Infof("invalid process %x status %s for setting the results, skipping",
+			results.ProcessID, vocProcessData.Status.String())
```
Comment (mvdan) [inline chain/ethevents/events.go]: no need to call String if the verb is already %s

File: client/api.go [hunk 2]
```diff
@@ -404,27 +403,29 @@ func (c *Client) TestSendVotes(
 		}
 	}
 	votingElapsedTime := time.Since(start)
-	for {
-		for i, nullHex := range nullifiers {
-			if nullHex == "registered" {
-				registered++
-				continue
+	if checkNullifiers {
```
Comment (mvdan) [inline client/api.go]: invert and early return to unindent the rest

File: cmd/vochaintest/vochaintest.go [hunk 3]
```diff
@@ -45,6 +45,7 @@ func main() {
 	parallelCons := flag.Int("parallelCons", 1, "parallel API connections")
 	procDuration := flag.Int("processDuration", 500, "voting process duration in blocks")
 	doubleVote := flag.Bool("doubleVote", true, "send every vote twice")
+	checkNullifiers := flag.Bool("checkNullifiers", false, "check all votes are correct one by one (slow)")
```
Comment (mvdan) [inline cmd/vochaintest/vochaintest.go:48]: So, turning this on is slower but... more secure? Or more reliable? I'm not sure if I get why anyone would ever want to turn it on.

If there's no reason to turn it on, I'm inclined to just not have this feature at all. And if the reason to turn it on is better security/reliability, I wonder why it's not on always.

File: vochain/state.go [hunk 4]
```diff
@@ -45,17 +45,25 @@ var PrefixDBCacheSize = 0
 
 // EventListener is an interface used for executing custom functions during the
 // events of the block creation process.
-// The order in which events are executed is: Rollback, OnVote or OnProcess, Commit.
+// The order in which events are executed is: Rollback, OnVote, Onprocess, On..., Commit.
 // The process is concurrency safe, meaning that there cannot be two sequences
 // happening in parallel.
+//
+// If Commit() returns an error and isFatal set tu true, it is considered a consensus
+// failure and the blockchain will halt.
```
Comment (mvdan) [inline vochain/state.go:53]: typo: "tu true"

I also think that just an `error` return would be simpler and clearer here. You could then use a special error type to specify fatal errors. For example:

```
type ErrHaltVochain struct {
    reason error
}

func (e ErrHaltVochain) Error() string { return fmt.Sprintf("halting vochain: %v", e.reason) }
func (e ErrHaltVochain) Unwrap() error { return e.reason }
```

Then, instead of:

```
err, fatal := l.Commit(h.Height)
if err != nil {
	if fatal {
		panic(err)
	}
	log.Warn(err)
}
```

you could do

```
err := l.Commit(h.Height)
if err != nil {
	if _, fatal := err.(ErrHaltVochain); fatal {
		panic(err)
	}
	log.Warn(err)
}
```

Or, even simpler, just panic directly in Commit if there's a fatal error. Having the caller panic is really the same, but more complex.

File: vochain/state.go [hunk 5]
```diff
@@ -100,19 +110,24 @@ func NewState(dataDir string) (*State, error) {
...
 	return vs, nil
 }
 
 func initStore(dataDir string, state *State) error {
+	log.Infof("initializing state db store")
 	state.Store = new(iavlstate.IavlState)
 	if err := state.Store.Init(dataDir, "disk"); err != nil {
 		return err
 	}
+	timer := time.Now()
```
Comment (mvdan) [inline vochain/state.go]: this is not a timer, so maybe call it `startTime`

File: vochain/state.go [hunk 6]
```diff
@@ -122,6 +137,7 @@ func initStore(dataDir string, state *State) error {
 	if err := state.Store.AddTree(VoteTree); err != nil {
 		return err
 	}
+	log.Infof("state tree load took %d ms", time.Since(timer).Milliseconds())
```
Comment (mvdan) [inline vochain/state.go]: I'd just use `%s` with `time.Since(startTime)`, it will already do the right thing by printing e.g. `200ms`. It will also behave better if the duration is over 1s.

File: vochain/scrutinizer/db.go [hunk 7]
```diff
@@ -47,9 +47,25 @@ type Entity struct {
...
 	ProcessID  types.HexBytes `badgerholdKey:"ProcessID"`
-	Votes      []*models.QuestionResult
+	Votes      [][]*big.Int
 	Weight     *big.Int
 	Signatures []types.HexBytes
+	Final      bool
+}
+
+func (r *Results) String() string {
+	results := ""
```
Comment (mvdan) [inline vochain/scrutinizer/db.go]: use a `strings.Builder`; otherwise you're allocating and copying the entire string at ever `+=` statement. Since there could be thousands of votes, this could be *very* slow, and indeed quadratic.

File: vochain/scrutinizer/scrutinizer.go [hunk 8]
```diff
@@ -34,16 +38,19 @@ func (s *Scrutinizer) AddEventListener(l EventListener) {
...
 	// updateProcessPool is the list of process IDs that require sync with the state database
 	updateProcessPool [][]byte
 	// resultsPool is the list of processes that finish on the current block
 	resultsPool []*types.ScrutinizerOnProcessData
+	// list of live processes (those on which the votes will be computed on arrival)
+	liveResultsProcs sync.Map
 	// eventListeners is the list of external callbacks that will be executed by the scrutinizer
 	eventListeners []EventListener
 	db             *badgerhold.Store
+	addVoteLock    sync.RWMutex
```
Comment (mvdan) [inline vochain/scrutinizer/scrutinizer.go]: please document what this mutex is actually guarding.

Comment (mvdan) [reply vochain/scrutinizer/scrutinizer.go]: For example, right now I don't get it - the DB is already atomic and thread-safe, so having a mutex around inserts/gets of votes seems entirely unnecessary. And if this is preventing data races, then we probably have a bug somewhere.

File: vochain/scrutinizer/scrutinizer.go [hunk 9]
```diff
@@ -56,52 +63,103 @@ func NewScrutinizer(dbPath string, state *vochain.State) (*Scrutinizer, error) {
 		return nil, err
 	}
 	s.VochainState.AddEventListener(s)
+	go s.AfterSyncBootstrap()
 	return s, nil
 }
 
+func (s *Scrutinizer) AfterSyncBootstrap() {
+	// Some grace time to avoid false positive on IsSynchronizing()
```
Comment (mvdan) [inline vochain/scrutinizer/scrutinizer.go]: I don't get this...

File: vochain/scrutinizer/scrutinizer.go [hunk 10]
```diff
@@ -56,52 +63,103 @@ func NewScrutinizer(dbPath string, state *vochain.State) (*Scrutinizer, error) {
...
+	time.Sleep(time.Second * 10)
+	for s.VochainState.IsSynchronizing() {
+		time.Sleep(time.Second * 2)
+	}
```
Comment (mvdan) [inline vochain/scrutinizer/scrutinizer.go:100]: is there no way to block until the state is synchronized? that would prevent the polling/sleeping, and supposedly also fix the 10s sleep at the beginning too.

For example, have one channel that is closed when the state is fully synced. Then, here we just need to `<-s.VochainState.Synced()` or something like that. When the channel is closed, that receive will unblock.

File: vochain/scrutinizer/scrutinizer.go [hunk 11]
```diff
@@ -56,52 +63,103 @@ func NewScrutinizer(dbPath string, state *vochain.State) (*Scrutinizer, error) {
...
-	if nvotes > 0 {
-		log.Infof("added %d live votes from block %d", nvotes, height)
 	}
 
 	// Update existing processes
 	for _, p := range s.updateProcessPool {
 		if err := s.updateProcess(p); err != nil {
-			log.Warnf("commit: cannot update process %x: %v", p, err)
+			log.Errorf("commit: cannot update process %x: %v", p, err)
+			continue
```
Comment (mvdan) [inline vochain/scrutinizer/scrutinizer.go]: this loop does nothing else, so I guess you can remove the continue

File: vochain/scrutinizer/scrutinizer.go [hunk 12]
```diff
@@ -56,52 +63,103 @@ func NewScrutinizer(dbPath string, state *vochain.State) (*Scrutinizer, error) {
...
+					nvotes++
+				}
+			}
+			if err := s.commitVotes([]byte(pid), results); err != nil {
+				log.Errorf("cannot commit live votes: (%v)", err)
+			}
+		}
+		if nvotes > 0 {
+			log.Infof("added %d live votes on block %d, took %d ms",
+				nvotes, height, time.Since(timer).Milliseconds())
```
Comment (mvdan) [inline vochain/scrutinizer/scrutinizer.go]: same comments about `timer` and `%s` like before :)

File: vochain/scrutinizer/scrutinizer.go [hunk 13]
```diff
@@ -56,52 +63,103 @@ func NewScrutinizer(dbPath string, state *vochain.State) (*Scrutinizer, error) {
...
+	// Add votes collected by onVote (live results)
+	// this can be run async
+	go func() {
+		nvotes := 0
+		timer := time.Now()
+		for pid, votes := range s.votePool {
+			results := &Results{Weight: new(big.Int).SetUint64(0)}
+			for _, v := range votes {
+				if err := s.addLiveVote(v, results); err != nil {
+					log.Warnf("vote cannot be added: %v", err)
```
Comment (mvdan) [inline vochain/scrutinizer/scrutinizer.go:215]: shouldn't this be Errorf?

File: vochain/scrutinizer/scrutinizer.go [hunk 14]
```diff
@@ -115,7 +173,11 @@ func (s *Scrutinizer) OnProcess(pid, eid []byte, censusRoot, censusURI string) {
 
 // OnVote scrutinizer stores the votes if liveResults enabled
 func (s *Scrutinizer) OnVote(v *models.Vote) {
-	s.votePool = append(s.votePool, v)
+	if _, ok := s.votePool[string(v.ProcessId)]; ok {
+		s.votePool[string(v.ProcessId)] = append(s.votePool[string(v.ProcessId)], v)
```
Comment (mvdan) [inline vochain/scrutinizer/scrutinizer.go:250]: I think you can just have this line. Because if the map key doesn't exist, you'll do `append` with the first argument being nil, which will work just fine.

File: vochain/scrutinizer/vote.go [hunk 15]
```diff
@@ -163,80 +135,125 @@ func (s *Scrutinizer) addLiveVote(envelope *models.Vote) error {
...
+	s.addVoteLock.Lock()
+	defer s.addVoteLock.Unlock()
+	update := func(record interface{}) error {
+		update, ok := record.(*Results)
+		if !ok {
+			return fmt.Errorf("record isn't the correct type! Wanted Result, got %T", record)
+		}
+		update.Weight.Add(update.Weight, results.Weight)
+
+		if len(results.Votes) > 0 {
```
Comment (mvdan) [inline vochain/scrutinizer/vote.go]: invert and early return to unindent

File: vochain/scrutinizer/vote.go [hunk 16]
```diff
@@ -163,80 +135,125 @@ func (s *Scrutinizer) addLiveVote(envelope *models.Vote) error {
...
+	var err error
+	// If badger returns "Transaction Conflict" we try again until it works
+	maxTries := 1000
+	for maxTries > 0 {
+		err = s.db.UpdateMatching(&Results{},
+			badgerhold.Where("ProcessID").Eq(pid).And("Final").Eq(false), update)
+		if err == nil {
+			break
+		}
+		if strings.Contains(err.Error(), "Transaction Conflict") {
```
Comment (mvdan) [inline vochain/scrutinizer/vote.go]: use badger.ErrConflict instead: https://pkg.go.dev/github.com/dgraph-io/badger/v2#pkg-variables

If the error is wrapped, then use `errors.Is(err, badger.ErrConflict)`.

File: vochain/scrutinizer/vote.go [hunk 17]
```diff
@@ -163,80 +135,125 @@ func (s *Scrutinizer) addLiveVote(envelope *models.Vote) error {
...
-	return &models.ProcessResult{
-		Votes:     results.Votes,
-		ProcessId: processID,
-	}, nil
+	if maxTries == 0 {
+		err = fmt.Errorf("got too much transaction conflicts")
+	}
+	if err != nil {
+		log.Debugf("saved %d votes with total weight of %s on process %x", len(results.Votes),
+			results.Weight.String(), pid)
```
Comment (mvdan) [inline vochain/scrutinizer/vote.go]: no need to call String with %s

**Review by mvdan — COMMENTED** [review]
I can't really approve because I fail to properly understand half of the changes :) If you have the time and want me to look more carefully, maybe we can do a 30m call and I'll ask questions while reading the changes.

File: vochain/scrutinizer/scrutinizer.go (hunk 13 above)
Comment (p4u) [reply vochain/scrutinizer/scrutinizer.go:215]: Warning is fine since the user might provide a wrong Vote package and that should be considered as a "null vote", it's not an error and probably in the future (when we are sure everything works as expected) we can even remove the warning

File: vochain/scrutinizer/vote.go (hunk 16 above)
Comment (p4u) [reply vochain/scrutinizer/vote.go]: Mmm, I did not add this error cause I did not want to import a specific KV implementation. So badgerhold can switch from badger v2 to v3 without any change in our side. Do you really think it is worth it? (importing badger v2 just for handling this error).

File: vochain/scrutinizer/scrutinizer.go (hunk 8 above)
Comment (p4u) [reply vochain/scrutinizer/scrutinizer.go]: That's for avoiding the "Transaction Conflict" error which happens quite often. That's a non critical error and the badger developers recommend to try it again until the db transaction works. But I preferred to try to avoid directly the put/get conflict with a Mutex. And yes, with that Mutex the iteration over the "Transaction Conflict" error might be removed but I let it there for security since I'm still not 100% sure all conflicting calls are mutex.

File: vochain/scrutinizer/scrutinizer.go (hunk 9 above)
Comment (p4u) [reply vochain/scrutinizer/scrutinizer.go]: During the first seconds/milliseconds of the Vochain startup, Tendermint might report that the chain is not synchronizing since it still does not have any peer and do not know the actual size of the blockchain. If AfterSyncBootStrap is executed on this specific moment, the Wait loop would pass.

File: vochain/scrutinizer/scrutinizer.go (hunk 10 above)
Comment (p4u) [reply vochain/scrutinizer/scrutinizer.go:100]: With the channel we would still have the same problem.

Maye a more safe and dynamic approach would be something like:
```
        syncSignal := 5
	for {
	is !s.VochainState.IsSynchronizing() { syncSignal-- }
	if syncSignal == 0 { break }
        time.Sleep(time.Second * 1)
	}
```

**Comment by p4u** [comment]
> I can't really approve because I fail to properly understand half of the changes :) If you have the time and want me to look more carefully, maybe we can do a 30m call and I'll ask questions while reading the changes.

Well I suppose you read the commit message already. I understand the changes are huge and there are plenty of design decisions on there kind of hidden for a third party developer. I have time for a short call this morning, else until 6th I won't be available.

File: vochain/scrutinizer/scrutinizer.go (hunk 8 above)
Comment (mvdan) [reply vochain/scrutinizer/scrutinizer.go]: If retrying works, then why have the mutex? I don't know, having a mutex around the entire DB feels hacky. If we have good tests and they continuously pass without the mutex and the error retry, then I'd say that's good.

File: vochain/scrutinizer/scrutinizer.go (hunk 9 above)
Comment (mvdan) [reply vochain/scrutinizer/scrutinizer.go]: Gotcha, maybe summarize that here?

File: vochain/scrutinizer/vote.go (hunk 16 above)
Comment (mvdan) [reply vochain/scrutinizer/vote.go]: Right, I wasn't thinking about v2 vs v3. If we have good tests which cover the conflict case (e.g. with many reads/writes), then I imagine the tests would catch if they bump to v3 but our error points at v2.

Alternatively, add a short comment to explain the strings.Contains? Most times it's a mistake, so it would be good to clarify why it's wanted here.

**Comment by mvdan** [comment]
No worries. Yeah, I read the commits and they make sense to me, though the code is kinda hard to follow. I think overall LGTM, the only major thing that worries me is why we want the checkNullifiers flag - left a comment about that earlier.

File: cmd/vochaintest/vochaintest.go (hunk 3 above)
Comment (p4u) [reply cmd/vochaintest/vochaintest.go:48]: This is a testing tool. I think the full check of nullifiers could be useful depending on the purpose of the testing. I will keep it available, it's not a big chunk of code.

File: vochain/state.go (hunk 4 above)
Comment (p4u) [reply vochain/state.go:53]: Great solution. Thanks.

**Comment by p4u** [comment]
Thanks for the review. Some day I'd like to have a "mvdan-linter" to include on golangci-lint :)

File: vochain/scrutinizer/scrutinizer.go (hunk 8 above)
Comment (p4u) [reply vochain/scrutinizer/scrutinizer.go]: Well, to me it looks that repeating a DB transaction N times until it success is kind of hacky. Using a Mutex on this scope makes sense and will release load on the KV context. 

In addition I added a `maxRetries=1000` mechanism which is kind of arbitrary (I tested different values until there were no more transactions failing). This value works in my local computer but a slower one could need more maxRetries. I don't like to block the goroutine forever if there is an issue on the KV (that won't accept new transactions) I prefer instead to try N times and raise an Error. So this is why I'm mixing both approaches: Mutex + Repeat Transaction N times

I'll add a to-do for removing it in the future, with proper research and testing involved. But for now I prefer to make it safe.
